LoRaEngine/modules/LoRaWanNetworkSrvModule/Logger/IotHubLogger.cs

```diff
@@ -0,0 +1,114 @@
+    {
+        private readonly ConcurrentDictionary<string, IotHubLogger> loggers = new();
+        private readonly Lazy<Task<ModuleClient>> moduleClientFactory;
+
+        public IotHubLoggerConfiguration Configuration { get; }
+        public LoggerExternalScopeProvider? ScopeProvider { get; }
+
+        public IotHubLoggerProvider(IotHubLoggerConfiguration configuration)
+            : this(configuration, new Lazy<Task<ModuleClient>>(ModuleClient.CreateFromEnvironmentAsync(new[] { new AmqpTransportSettings(TransportType.Amqp_Tcp_Only) })))
+        { }
+
+        internal IotHubLoggerProvider(IotHubLoggerConfiguration configuration, Lazy<Task<ModuleClient>> moduleClientFactory)
+        {
+            Configuration = configuration;
+            ScopeProvider = configuration.UseScopes ? new LoggerExternalScopeProvider() : null;
+            this.moduleClientFactory = moduleClientFactory;
+        }
+
+        public ILogger CreateLogger(string categoryName) =>
+            this.loggers.GetOrAdd(categoryName, n => new IotHubLogger(this, this.moduleClientFactory));
+
+        public void Dispose()
+        {
+            this.loggers.Clear();
+        }
+    }
+
+    internal class IotHubLogger : ILogger
+    {
+        private readonly IotHubLoggerProvider iotHubLoggerProvider;
+        private readonly Lazy<Task<ModuleClient>> moduleClientFactory;
+
+        public IotHubLogger(IotHubLoggerProvider iotHubLoggerProvider,
+                            Lazy<Task<ModuleClient>> moduleClientFactory)
+        {
+            this.iotHubLoggerProvider = iotHubLoggerProvider;
+            this.moduleClientFactory = moduleClientFactory;
+        }
+
+        public IDisposable BeginScope<TState>(TState state) =>
+            this.iotHubLoggerProvider.ScopeProvider?.Push(state) ?? NoopDisposable.Instance;
+
+        public bool IsEnabled(LogLevel logLevel) =>
+            logLevel >= this.iotHubLoggerProvider.Configuration.LogLevel;
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
+        {
+            _ = formatter ?? throw new ArgumentNullException(nameof(formatter));
+
+            if (!IsEnabled(logLevel))
+                return;
+
+            var configuredEventId = this.iotHubLoggerProvider.Configuration.EventId;
+            if (configuredEventId == 0 || configuredEventId == eventId)
+            {
+                _ = Task.Run(async () =>
+                {
+                    try
+                    {
+                        var formattedMessage = LoggerHelper.AddScopeInformation(this.iotHubLoggerProvider.ScopeProvider, formatter(state, exception));
+                        await SendAsync(formattedMessage);
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine($"Error when sending log to IoT Hub: {ex}");
+                        throw;
```

Comment:
this is not doing much, right? The throw will go un-noticed. Not sure we need to re-throw here? Depending on the error, would we ever want to stop trying to send the messages? Maybe we can continue to send when we encounter an IoT hub exception that we consider transient and stop logging for other exceptions? Seems like a pretty expensive path, if we continuously fail?

Reply:
Good point about the "permanently broken" logger, I will see what exceptions I can provoke and when I can "disable" the logger.

I don't think we should hide the exception here, though:
- It will generate a CA warning that we need to suppress
- It should be up to the caller to decide how to handle unhandled exceptions. If we ever decide to make use of something like `TaskScheduler.UnobservedTaskException`, we will not get these exceptions without knowing that we hide the exceptions here

Reply:
Did the first part in 4865a49dedb2289bbfb785cd1fdf9f847286807f